thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp

```diff
@@ -45,1910 +49,2237 @@
-            memcpy(fetchInRowPtr, &fpos, sizeof(fpos));
-            fetchInRowPtr += sizeof(fpos); 
-            memcpy(fetchInRowPtr, &jg, sizeof(jg));
-            fetchInRowPtr += sizeof(jg);
-
-            size32_t sz = 0;
-            if (owner.fetchInputAllocator)
+            unsigned partCopy = parts[selected];
+            unsigned partNo = partCopy & partMask;
+            unsigned copy = partCopy >> 24;
+            IKeyManager *&keyManager = keyManagers[selected];
+            if (!keyManager) // delayed until actually needed
             {
-                RtlDynamicRowBuilder fetchRow(owner.fetchInputAllocator);
-                sz = owner.helper->extractFetchFields(fetchRow, jg->queryLeft());
-                const void *fFetchRow = fetchRow.finalizeRowClear(sz);
-                memcpy(fetchInRowPtr, &fFetchRow, sizeof(const void *));  // child row of fetchInRow
+                keyManager = activity.createPartKeyManager(partNo, copy);
+                // NB: potentially translation per part could be different if dealing with superkeys
+                setupTranslation(partNo, *keyManager);
             }
+            processRows(processing, partNo, keyManager);
+        }
+    };
+    class CKeyLookupMergeHandler : public CKeyLookupLocalBase
+    {
+        typedef CKeyLookupLocalBase PARENT;
 
-            if (totalSz + (FETCHKEY_HEADER_SIZE+sz) > minFetchSendSz)
-            {
-                sendAll(); // in effect send remaining
-                totalSz = 0;
-            }
-            unsigned dstNode;
-            if (owner.remoteDataFiles)
-                dstNode = owner.node; // JCSMORE - do directly
-            else
+        Owned<IKeyManager> keyManager;
+    public:
+        CKeyLookupMergeHandler(CKeyedJoinSlave &_activity) : CKeyLookupLocalBase(_activity)
+        {
+            limiter = &activity.lookupThreadLimiter;
+        }
+        virtual void process(CThorExpandingRowArray &processing, unsigned __unused) override
+        {
+            if (!keyManager)
             {
-                if (1 == owner.filePartTotal)
-                    dstNode = owner.globalFPosToNodeMap[0].index;
-                else if (isLocalFpos(fpos))
-                    dstNode = getLocalFposPart(fpos);
-                else
+                Owned<IKeyIndexSet> partKeySet = createKeyIndexSet();
+                for (auto &partCopy: parts)
                 {
-                    const void *result = bsearch(&fpos, owner.globalFPosToNodeMap, owner.filePartTotal, sizeof(FPosTableEntry), slaveLookup);
-                    if (!result)
-                        throw MakeThorException(TE_FetchOutOfRange, "FETCH: Offset not found in offset table; fpos=%" I64F "d", fpos);
-                    dstNode = ((FPosTableEntry *)result)->index;
+                    unsigned partNo = partCopy & partMask;
+                    unsigned copy = partCopy >> 24;
+                    Owned<IKeyIndex> keyIndex = activit
```

Comment:
Use ScopedAtomic in jatomic.hpp to avoid an atomic inc for each iteration around the loop.

Anywhere an atomic is being iterated in a loop it is likely to be more efficient to use ScopedAtomic - it it exception safe.  Only complication is you would need to add the local to the stat in the line below.

Reply:
line 1233

Reply:
(and following)

Reply:
>1233:                    activity.statsArr[AS_DiskAccepted]++;

Ok, sounds good, will change to use ScopedAtomics.